library/proc_macro/src/quote.rs

```diff
@@ -477,6 +407,381 @@ fn collect_meta_vars(content_stream: TokenStream) -> Vec<Ident> {
     vars
 }
 
+/// Consume a `$( CONTENTS ) SEP_OPT *` accordingly.
+fn consume_dollar_group_sep_star(
+    content_stream: TokenStream,
+    iter: &mut Peekable<crate::token_stream::IntoIter>,
+) -> TokenStream {
+    let mut tokens = crate::TokenStream::new();
+
+    let mut current_contents: TokenStream = content_stream;
+    let mut sep_cand = Vec::new();
+    let mut is_sep_confirmed = false;
+    loop {
+        match (iter.next(), iter.peek().cloned()) {
+            // If a valid `*` is found, expand the current_contents and consume the `*`.
+            (Some(TokenTree::Punct(star)), opt)
+                if star.as_char() == '*' && !matches!(opt, Some(TokenTree::Punct(_))) =>
+            {
+                let sep_opt: Option<TokenStream> =
+                    (!sep_cand.is_empty()).then(|| sep_cand.into_iter().collect::<TokenStream>());
+
+                expand_dollar_group_sep_star(current_contents.clone(), sep_opt)
+                    .to_tokens(&mut tokens);
+                break;
+            }
+            (Some(TokenTree::Punct(star)), Some(TokenTree::Punct(not_assign)))
+                if star.as_char() == '*'
+                    && !(star.spacing() == Spacing::Joint && not_assign.as_char() == '=') =>
+            {
+                let sep_opt: Option<TokenStream> =
+                    (!sep_cand.is_empty()).then(|| sep_cand.into_iter().collect::<TokenStream>());
+
+                expand_dollar_group_sep_star(current_contents.clone(), sep_opt)
+                    .to_tokens(&mut tokens);
+                break;
+            }
+
+            // If the next `$( CONTENTS )` is found before the `*`, consume the current `$( CONTENTS )` literally.
+            // Then move to consume the next `$( CONTENTS )`.
```

Comment:
I do not recognize what case this arm is trying to handle. I also was unable to come up with any `quote!` invocation where this arm would expand to something that even compiles. Please add a test that covers what is being done here.